### PR TITLE
refactor: consolidate store-state nil-checking

### PR DESCRIPTION
**Pull request by Shou**
We make the nil-checking of store state fields generic through a
`verifyNoNilFields` function that throws an error if any fields are
nil.

Change-Type: patch
Changelog-Entry: Consolidate store state nil-checking with helper
function.

### CODE REVIEW

File: lib/shared/store.js [hunk 1]
```diff
@@ -176,48 +201,32 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
         })
       }
 
-      if (!action.data.type) {
-        throw errors.createError({
-          title: 'Missing state type'
-        })
-      }
+      const fields = [
```
Comment (jhermsmeier) [inline lib/shared/store.js]: This should be a constant way higher up, out of the local scope; we wouldn't want an array to be allocated for this on each call

File: lib/shared/store.js [hunk 2]
```diff
@@ -319,24 +328,19 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
     // place where all the image extension / format handling
     // takes place, to avoid having to check 2+ locations with different logic
     case ACTIONS.SELECT_IMAGE: {
-      if (!action.data.path) {
-        throw errors.createError({
-          title: 'Missing image path'
-        })
-      }
+      const noNilFields = [
```
Comment (jhermsmeier) [inline lib/shared/store.js]: Same here – let's try to avoid allocating things that are constant over and over

File: lib/shared/store.js [hunk 3]
```diff
@@ -28,6 +28,31 @@ const fileExtensions = require('./file-extensions')
...
+ * const fields = [ 'type', 'percentage' ]
+ * verifyNoNilFields(action.data, fields)
+ */
+const verifyNoNilFields = (object, fields, name) => {
+  const nilFields = _.filter(fields, (field) => {
+    return _.isNil(_.get(object, field))
+  })
+  if (nilFields.length) {
+    throw errors.createError({
+      title: `Missing ${name} fields: ${nilFields.join(', ')}`
```
Comment (jhermsmeier) [inline lib/shared/store.js]: We can just use ```throw new Error(`Missing ${name} fields: ${nilFields.join(', ')}`)``` here

**Comment by Shou** [comment]
@jhermsmeier done – moved the constants to the global scope and removed `errors` usage

File: lib/shared/store.js [hunk 4]
```diff
@@ -28,6 +28,52 @@ const fileExtensions = require('./file-extensions')
...
+ * @public
+ *
+ * @param {Object} object - state object
+ * @param {Array<Array<String>> | Array<String>} fields - array of object field paths
+ * @param {String} name - name of the state we're dealing with
+ * @throws
+ *
+ * @example
+ * const fields = [ 'type', 'percentage' ]
+ * verifyNoNilFields(action.data, fields)
```
Comment (lurch) [inline lib/shared/store.js]: Should the example show the `name` parameter being used?

File: lib/shared/store.js [hunk 5]
```diff
@@ -28,6 +28,52 @@ const fileExtensions = require('./file-extensions')
 const utils = require('./utils')
 const packageJSON = require('../../package.json')
 
+/**
+ * @summary Verify and throw if any state fields are nil
+ * @function
+ * @public
+ *
+ * @param {Object} object - state object
+ * @param {Array<Array<String>> | Array<String>} fields - array of object field paths
```
Comment (lurch) [inline lib/shared/store.js]: Is this line correct? Should it just be
```
 * @param {Array<String>} fields - array of object field paths
```
?

File: lib/shared/store.js [hunk 6]
```diff
@@ -28,6 +28,52 @@ const fileExtensions = require('./file-extensions')
...
+ * @param {String} name - name of the state we're dealing with
+ * @throws
+ *
+ * @example
+ * const fields = [ 'type', 'percentage' ]
+ * verifyNoNilFields(action.data, fields)
+ */
+const verifyNoNilFields = (object, fields, name) => {
```
Comment (lurch) [inline lib/shared/store.js]: Is it worth creating a standalone unit-test for this function? Or are the existing unit-tests sufficient?
(I note for example that there's nothing testing that multiple fields are missing at once, which I guess means the `.join(', ')` part isn't being tested?)

File: lib/shared/store.js (hunk 5 above)
Comment (Shou) [reply lib/shared/store.js]: It's correct, it just passes its argument to `_.get` which can take a string or an array of string for more complex path traversal.

File: lib/shared/store.js (hunk 4 above)
Comment (Shou) [reply lib/shared/store.js]: Will submit a PR for this

File: lib/shared/store.js (hunk 6 above)
Comment (Shou) [reply lib/shared/store.js]: That's true, I'll add tests for multiple missing fields.

File: lib/shared/store.js (hunk 5 above)
Comment (lurch) [reply lib/shared/store.js]: Ahh, then perhaps that behaviour ought to be unit-tested too then?

**Pull request merged**